cmake/ChooseBLAS.cmake

```diff
@@ -0,0 +1,266 @@
+    return()
+  endif()
+
+  if(NOT SYSTEM_ARCHITECTURE STREQUAL "x86_64")
+    message(WARNING "MKL is supported only for desktop platforms (SYSTEM_ARCHITECTURE=${SYSTEM_ARCHITECTURE}), \
+                     MKL will not be available")
+    return()
+  endif()
+
+  find_package(MKL)
+
+  if(MKL_FOUND)
+    message(STATUS "MKL framework found")
+
+    set(MKL_FOUND ${MKL_FOUND} PARENT_SCOPE)
+    set(MKL_INCLUDE_DIR ${MKL_INCLUDE_DIR} PARENT_SCOPE)
+    set(MKL_LIBRARIES ${MKL_LIBRARIES} PARENT_SCOPE)
+    set(MKLROOT ${MKLROOT} PARENT_SCOPE)
+
+    set(BLAS MKL PARENT_SCOPE)
+  else()
+    message(STATUS "MKL framework not found")
+  endif()
+
+endfunction()
+
+function(try_mklml)
+  if(NOT USE_MKLML)
+    return()
+  endif()
+
+  if(MKL_FOUND)
+    return()
+  endif()
+
+  message(STATUS "Trying to enable MKLML framework due to \
+                  USE_MKLML=${USE_MKLML} and USE_MKL_IF_AVAILABLE=${USE_MKL_IF_AVAILABLE}")
+
+  if(CMAKE_CROSSCOMPILING)
+    message(WARNING "MKLML with cross compilation is not supported, MKL will not be available")
+    return()
+  endif()
+
+  if(NOT SYSTEM_ARCHITECTURE STREQUAL "x86_64")
+    message(WARNING "MKL is supported only for desktop platforms (SYSTEM_ARCHITECTURE=${SYSTEM_ARCHITECTURE}), \
+                     MKL will not be available")
+    return()
+  endif()
+
+  include(${CMAKE_CURRENT_LIST_DIR}/DownloadMKLML.cmake)
+  find_package(MKLML REQUIRED)
+
+  set(MKL_FOUND ${MKL_FOUND} PARENT_SCOPE)
+  set(MKL_INCLUDE_DIR ${MKL_INCLUDE_DIR} PARENT_SCOPE)
+  set(MKL_LIBRARIES ${MKL_LIBRARIES} PARENT_SCOPE)
+  set(MKLROOT ${MKLROOT} PARENT_SCOPE)
+
+  set(BLAS MKL PARENT_SCOPE)
+
+  message(STATUS "MKLML framework found")
+
+endfunction()
+
+function(try_accelerate)
+  if(NOT APPLE)
+    return()
+  endif()
+
+  if(BLAS MATCHES "[Mm][Kk][Ll]")
+    return()
+  endif()
+
+  if(USE_APPLE_ACCELERATE_IF_AVAILABLE)
+    message(STATUS "Trying to enable Apple Accelerate framework due to USE_ACCELERATE_IF_AVAILABLE")
+    find_package(Accelerate)
+    if(Accelerate_FOUND)
+      message(STATUS "Apple Accelerate framework found")
+      set(BLAS Accelerate PARENT_SCOPE)
+    else()
+      message(STATUS "Apple Accelerate framework not found")
+    endif()
+  endif()
+endfunction()
+
+if(USE_MKL_IF_AVAILABLE)
+  set(MKL_FOUND)
+
+  try_mkl()
+  try_mklml()
+  try_mkldnn()
+endif()
+
+try_accelerate()
+
+# cmake regexp does not support case insensitive match (?i) or //i
+if(BLAS MATCHES "[Aa][Tt][Ll][Aa][Ss]")
+  message(STATUS "Using Atlas for BLAS")
+
+  set(Atlas_NEED_LAPACK ${USE_LAPACK})
+  find_package(Atlas REQUIRED)
+
+  include_directories(SYSTEM ${Atlas_INCLUDE_DIRS})
+  set(mxnet_LINKER_LIBS ${mxnet_LINKER_LIBS} ${Atlas_LIBRARIES})
+
+  add_definitions(-DMSHADOW_USE_CBLAS=1)
+  add_definitions(-DMSHADOW_USE_MKL=0)
+
+  if(USE_LAPACK AND Atlas_LAPACK_FOUND)
+    switch_lapack(True)
+  else()
+    switch_lapack(False)
+  endif()
+
+  return()
+endif()
+
+if(BLAS MATCHES "[Oo][Pp][Ee][Nn]")
```

Comment:
could we use TOLOWER? https://cmake.org/pipermail/cmake/2012-June/050651.html
this takes more time to read (and looks funny)

Reply:
Done